### PR TITLE
fixes #24: ES6: arrow functions

### LINKED ISSUE #24: ES6: arrow functions
# [14.2](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-arrow-function-definitions) Arrow Function Definitions

### Syntax

```
ArrowFunction[In, Yield] :
  ArrowParameters[?Yield] [no LineTerminator here] => ConciseBody[?In]
ArrowParameters[Yield] :
  BindingIdentifier[?Yield]
  CoverParenthesizedExpressionAndArrowParameterList[?Yield]
ConciseBody[In] :
  [lookahead ≠ { ] AssignmentExpression[?In]
  { FunctionBody }
```

### Supplemental Syntax

When the production

    ArrowParameters[Yield] : CoverParenthesizedExpressionAndArrowParameterList[?Yield]

is recognized the following grammar is used to refine the interpretation of `CoverParenthesizedExpressionAndArrowParameterList`:

```
ArrowFormalParameters[Yield, GeneratorParameter] :
  ( StrictFormalParameters[?Yield, ?GeneratorParameter] )
```

### Early Errors

    ArrowFunction : ArrowParameters => ConciseBody

- It is a Syntax Error if any element of the BoundNames of ArrowParameters also occurs in the LexicallyDeclaredNames of ConciseBody.

    ArrowParameters[Yield] : CoverParenthesizedExpressionAndArrowParameterList[?Yield]

- If the [Yield] grammar parameter is present on ArrowParameters, it is a Syntax Error if the lexical token sequence matched by CoverParenthesizedExpressionAndArrowParameterList[?Yield] cannot be parsed with no tokens left over using ArrowFormalParameters[Yield, GeneratorParameter] as the goal symbol.

- If the [Yield] grammar parameter is not present on ArrowParameters, it is a Syntax Error if the lexical token sequence matched by CoverParenthesizedExpressionAndArrowParameterList[?Yield] cannot be parsed with no tokens left over using ArrowFormalParameters as the goal symbol.

- All early errors rules for ArrowFormalParameters and its derived productions also apply to CoveredFormalsList of CoverParenthesizedExpressionAndArrowParameterList[?Yield].

NOTE: The yield operator cannot be used within expressions that are part of an ArrowFormalParameters.

### PR DESCRIPTION
**Pull request by ikarienator**
Fixes #24: ES6: arrow functions

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,7 +24,7 @@
     "shift-ast": "git://github.com/shapesecurity/shift-ast-js.git#591c3376fde127c6bdf6d5c61c00ab309af456c3"
   },
   "devDependencies": {
-    "6to5": "^1.14.14",
+    "6to5": "^3.4.1",
```
Comment (michaelficarra) [inline package.json]: We will lock to specific versions from now on, to help ensure deterministic builds. Remove the caret and submit this change as a separate PR..

Comment (ikarienator) [inline package.json]: Would you merge it first then change the caret?

Comment (michaelficarra) [inline package.json]: I think this unrelated change should be a separate PR to begin with.

File: src/parser.js [hunk 2]
```diff
@@ -1364,11 +1406,103 @@ export class Parser extends Tokenizer {
...
+        type: ARROW_EXPRESSION_PARAMS,
+        params: [],
+        rest: rest
+      };
+    }
+
+    let possibleBindings = !this.match(TokenType.LPAREN);
+    let startTokenIndex = this.tokenIndex;
+    let params = [this.parseAssignmentExpression()];
+    let group = params[0];
```
Comment (michaelficarra) [inline src/parser.js]:
```js
let group = this.parseAssignmentExpression();
let params = [group];
```

File: src/parser.js [hunk 3]
```diff
@@ -32,8 +32,13 @@ const INIT_MASK = 1;
...
 const SETTER_MASK = 4;
 
-const STRICT_MODE_RESERVED_WORD = [
-  "implements", "interface", "package", "private", "protected", "public", "static", "yield", "let"];
+// Empty parameter list for ArrowExpression
+const ARROW_EXPRESSION_PARAMS = "CoverParenthesizedExpressionAndArrowParameterList";
+
+const STRICT_MODE_RESERVED_WORD = {
+  "implements": true, "interface": true, "package": true, "private": true, "protected": true,
+  "public": true, "static": true, "yield": true, "let": true
```
Comment (michaelficarra) [inline src/parser.js]: I opt to assign a falsey value to these kinds of maps in case someone is tempted to check existence via reguar member access. Also, we can use `__proto__: null` if we want to use an `in` test.

File: src/parser.js [hunk 4]
```diff
@@ -86,40 +91,27 @@ const BinaryPrecedence = {
...
-        return result;
-    } else {
-      ++cursorB;
-      if (cursorB >= stringsB.length)
-        return result;
-    }
-  } while(true);
-  throw new Error("intersection algorithm broken");
+function hasStrictModeReservedWord(ids) {
+  return ids.some(id => STRICT_MODE_RESERVED_WORD.hasOwnProperty(id));
```
Comment (michaelficarra) [inline src/parser.js]: Just in case, I always pull `hasOwnProperty` off `{}`.

File: src/parser.js [hunk 5]
```diff
@@ -178,7 +170,7 @@ export class Parser extends Tokenizer {
   }
 
   parseScript() {
-    var [body, isStrict] = this.parseBody(true);
+    let [body, isStrict] = this.parseBody(true);
```
Comment (michaelficarra) [inline src/parser.js]: `isStrict` is unused?

File: src/parser.js [hunk 6]
```diff
@@ -954,6 +949,53 @@ export class Parser extends Tokenizer {
 
     let node = this.parseConditionalExpression();
 
+    let arrow;
+    if (!this.hasLineTerminatorBeforeNext && (arrow = this.eat(TokenType.ARROW))) {
+      // Convert param list.
```
Comment (michaelficarra) [inline src/parser.js]: Can you extract this to a function?

File: src/parser.js [hunk 7]
```diff
@@ -1364,11 +1406,103 @@ export class Parser extends Tokenizer {
...
+      if (this.strict && strict_restricted_word) {
+        throw this.createError(ErrorMessages.STRICT_PARAM_NAME);
+      }
+
+      let strict_reserved_word = hasStrictModeReservedWord(allBoundNames);
+      if (this.strict && strict_reserved_word) {
+        throw this.createError(ErrorMessages.STRICT_RESERVED_WORD);
+      }
+
+      return {type: ARROW_EXPRESSION_PARAMS, params, rest, dup, strict_reserved_word, strict_restricted_word};
```
Comment (michaelficarra) [inline src/parser.js]: This line is too long.

Comment (michaelficarra) [inline src/parser.js]: Duplicate `strict_reserved_word`.

File: test/functions/arrow-expression.js [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    testEsprimaEquiv("(a)");
+    testEsprimaEquiv("((a))");
+    testEsprimaEquiv("((a))()");
+    testEsprimaEquiv("((a))((a))");
+    testEsprimaEquiv("void (a)");
+    testEsprimaEquiv("(void a)");
+    testEsprimaEquiv("(a++)");
+    testEsprimaEquiv("(a)++");
+    testEsprimaEquiv("(a)--");
+    testEsprimaEquiv("(a) ? (b) : (c)");
```
Comment (michaelficarra) [inline test/functions/arrow-expression.js]: We really shouldn't be adding new esprima-equiv tests. We're trying to cut these out eventually.

**Comment by michaelficarra** [comment]
I've rebased off `es6` for you. Please pull.

File: test/functions/arrow-expression.js [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+    testParse("(()=>0)", expr, new Shift.ArrowExpression([], null, new Shift.LiteralNumericExpression(0)));
```
Comment (michaelficarra) [inline test/functions/arrow-expression.js]: Put a newline between each `testParse` for us vim `{` and `}` users.

File: test/functions/arrow-expression.js (hunk 8 above)
Comment (ikarienator) [inline test/functions/arrow-expression.js]: That's very boring to write down tough :-1:
